### ISSUE TITLE
Perceptual Reflex Layer + Compost Translation — sensors without governor authority

### ISSUE THREAD

**Issue by the-static-collective** [the-static-collective/the-haunted-toaster#184]
## Summary

Give Haunted Toaster, and potentially the wider eCODEsystem, a reusable way to borrow open-source AI / ML / DSP / CV machinery as **senses and reflexes without turning the appliance into an AI governor**.

The core architectural distinction is:

> **A learned or foreign component may produce testimony. It may never produce authority.**

And for research-only dependencies:

> **Knowledge may cross; dependency stays behind.**

This issue introduces two related primitives:

1. **Perceptual Reflex Layer** — bounded local sensors emit typed observations; deterministic/native policy decides whether those observations may influence proposal space or later admitted execution.
2. **Compost Translation** — a foreign research organism may be instrumented, characterized, and progressively replaced by native eCODEsystem primitives. Its contribution remains attributable even after the dependency is dropped.

Essentia is the proposed **First Dropout** specimen for Compost Translation.

This is an architectural / research slice. It does not authorize adding a giant model stack, a supervisor agent, or renderer-local intelligence.

## Immediate ancestry

This should compose existing project law rather than create a parallel authority system.

- #183 already establishes **Specimen Pulse v1**: attached VSPantry audio may be heard as deterministic influence without entering the final song master.
- #44 owns foreign visual material and local deterministic clip analysis.
- #168 / PR #170 establish proposal-only Creative Context / Influence Diet boundaries.
- #148 / #147 establish candidate ecology, lineage, replayability, and no AI winner selection.
- Existing Ghost Topology / possession / renderer-trust issues retain their own execution authority.

The first grounded Toaster vertical slice should therefore remain aligned with #183 rather than widening renderer authority here.

## Core architecture

```text
source audio / video / lyrics / runtime state
                |
                v
        PERCEPTION ORGANS
    DSP / CV / frozen embeddings
    similarity / clustering / probes
                |
                v
        typed observations
        authority: testimony only
                |
                v
         REFLEX / REGULATOR
      deterministic native policy
       state machine / rules
                |
                v
       bounded proposal pressure
                |
                v
    existing constituted authority chain
VisualScore -> ResolvedTimeline -> preview/render -> receipt
```

A sensor never gets to decide that its observation is true enough to become execution authority. The receiving system decides how, whether, and at what stage the observation may matter.

## Observation contract

A useful generic sensor envelope is conceptually:

```text
SensorWitness {
  schema,
  sensorId,
  sensorVersion,
  implementationKind,
  sourceSha256,
  inputInterval,
  parameters,
  observation,
  confidence,
  observationHash,
  authority: "testimony-only",
  model?: {
    identity,
    version,
    sha256,
    licenseWitness
  }
}
```

Exact field naming belongs to implementation. The invariant is that a sensor's output is a witness with provenance, not ambient state.

## Candidate organ catalog

These are research candidates, not mandatory dependencies. Every adoption must preserve local-first behavior, deterministic/replayable semantics where technically possible, and exact dependency/model licensing evidence.

### 1. librosa — classical audio / MIR sensor bench

Likely uses:

- onset strength / transient evidence;
- beat / pulse hypotheses;
- spectral centroid / bandwidth / flux;
- energy envelopes;
- harmonic/percussive evidence;
- section-change / recurrence features.

Fit: strong research and production candidate for analysis-only work; Python/runtime packaging cost still matters.

Current upstream code license observed 2026-08-19: ISC.

### 2. OpenCV — visual geometry and motion senses

Likely uses:

- frame differencing;
- optical flow / motion vector fields;
- foreground/background change;
- edge / contour / occupancy measures;
- quiet-region / motion-pressure witnesses;
- deterministic image transforms.

Fit: strong classical CV reference and possible production component. Prefer low-level geometry before semantic interpretation.

Current OpenCV 4.5+ upstream license observed 2026-08-19: Apache-2.0.

### 3. DINOv2 — frozen visual feature geometry

Likely uses:

- visual embeddings for candidate-family distance;
- near-duplicate / collapse detection;
- novelty / coverage measurement;
- image retrieval / nearest-neighbor relationships;
- visual topology neighborhoods without taste ranking.

Fit: learned microscope. It may measure difference; it must not choose the winner.

Current upstream DINOv2 code/model license observed 2026-08-19: Apache-2.0 for the primary DINOv2 code and weights; exact downstream heads/assets still require their own license witness.

### 4. MediaPipe — anatomy / landmark / segmentation sensors

Likely uses:

- face/body/hand landmark geometry;
- foreground masks;
- pose axes;
- topology-aware typography avoidance;
- bounded body-attached visual behavior.

Fit: geometry source, not semantic personhood. Use landmarks/masks as measurements rather than emotional or identity claims.

Current upstream code license observed 2026-08-19: Apache-2.0. MediaPipe documentation currently states input processing is on-device, while product telemetry behavior must be inspected separately before adoption.

### 5. Sentence Transformers / small text embeddings — semantic coordinates without generation

Likely uses:

- lyric-line similarity;
- motif recurrence;
- specimen/tag proximity;
- candidate/context retrieval;
- semantic distance as one bounded proposal signal.

Fit: text microscope, not writer. Framework license does not imply any particular downloaded model has the same license; model identity/hash/license must be witnessed separately.

Current framework license observed 2026-08-19: Apache-2.0.

### 6. XState — deterministic reflex topology

Likely uses:

- explicit state machines / statecharts;
- sensor-pressure thresholds;
- refusal / degraded-sensor states;
- predictable orchestration of bounded reflexes;
- inspectable transition history.

Fit: strong JS-native candidate for regulation. Use deterministic state-machine semantics; do not import an AI-agent supervisor merely because the surrounding ecosystem also uses actor terminology.

Current upstream license observed 2026-08-19: MIT.

### 7. ONNX Runtime — model-cartridge socket

Likely uses:

- local inference for explicitly admitted frozen models;
- one boring adapter from tensor input to typed witness output;
- model/version/hash isolation from application authority.

Fit: useful if the ecosystem later needs a stable local inference boundary without carrying full training frameworks.

Current upstream license observed 2026-08-19: MIT. Every model loaded through it still needs independent model/data/license provenance.

### 8. BehaviorTree.CPP — robotics reflex research reference

Likely uses:

- reactive tree composition ideas;
- bounded action/refusal semantics;
- inspectable behavior transitions;
- design reference for more complex reflexes than a simple statechart.

Fit: primarily a research/reference organism for now. The Toaster is JS-native enough that copying the behavioral ideas into native architecture is likely preferable to adding a C++ runtime dependency.

Current upstream license observed 2026-08-19: MIT.

### 9. Essentia — First Dropout / research organism

Essentia spans classical DSP, statistical/spectral/temporal/tonal descriptors, beat/rhythm analysis, and optional learned-model wrappers. That breadth makes it unusually useful for answering:

> **What senses should this organism possess?**

It is not proposed as a casually shipped Toaster dependency.

Current upstream licensing observed 2026-08-19:

- Essentia library: AGPLv3 for non-commercial/open use, proprietary licensing available;
- pretrained MTG model terms are separate and must be read per exact model;
- Essentia's own current licensing/model pages are not perfectly aligned on blanket model wording, so exact model LICENSE + file hash must outrank memory or a generic table;
- some optional third-party dependencies carry their own GPL/LGPL/other terms.

This licensing complexity is part of why Essentia is a strong founding Compost Translation specimen.

## Compost Translation

The purpose is not to make an Essentia-compatible clone.

The purpose is to let foreign machinery be **digested into native eCODEsystem grammar**.

```text
foreign research organism
        |
      INGEST
        |
 controlled specimens + receipts
        |
      DIGEST
        |
 hypotheses / behavioral specification
        |
       GROW
        |
 native eCODEsystem primitive
        |
     COMPOST
        |
 foreign runtime dependency no longer required
```

### First Dropout law

A foreign tool may be present in the research stage while a native replacement is being characterized.

The thing allowed to graduate is understanding:

- measured behavior;
- public mathematics / papers / standards;
- independently stated behavioral specification;
- test corpus and tolerances;
- failures / unresolved ambiguity;
- native primitive semantics.

The following do **not** automatically graduate merely because receipts exist:

- copied source code;
- linked binaries;
- model weights;
- vendor/private implementation details;
- an inherited license obligation that the production project has not deliberately accepted.

Receipts prove lineage and separation; they are not a magic license eraser.

## Instrumented black-box characterization

For classical / behaviorally characterizable functions, the research harness should support controlled probe families such as:

```text
silence
single impulse
periodic impulses
sine tones
frequency sweeps
white / pink noise
amplitude ramps
multi-tone mixtures
sample-rate variants
frame/window-size variants
real fixed WAV specimens
```

Each reference execution should produce an attributable observation:

```text
ReferenceObservation {
  foreignToolVersion,
  algorithm,
  parameters,
  inputSha256,
  inputGenerator,
  output,
  outputHash,
  licenseWitness
}
```

Hypotheses should themselves be versioned and allowed to fail:

```text
H17
  explains: R001-R438
  fails: R439
  supersedes: H11
```

The native implementation then earns graduation through differential tests against the behavioral specification and fixed specimens.

## Clean-room-friendly stage separation

Where independence matters, prefer an explicit process boundary:

```text
REFERENCE LAB
  may execute foreign tool
  may inspect documented API / public papers
  emits behavioral spec + receipts

             || only spec crosses ||

NATIVE IMPLEMENTATION
  consumes behavioral spec / public math
  does not import foreign runtime/source

             || differential proof ||

REFERENCE OUTPUT <-> NATIVE OUTPUT
```

The exact legal process should be reviewed if a future target involves patents, access-control circumvention, proprietary contracts, or model-output cloning. This issue does not declare every reverse-engineering technique lawful in every jurisdiction.

## Native growth rule

A graduating primitive should be reshaped around eCODEsystem needs instead of preserving a foreign library's API merely for compatibility.

Example:

```text
foreign concept:
  detectOnsets(audio)

native concept:
  witnessTransientPressure(window)
    -> interval
    -> pressure
    -> confidence
    -> contributing bands
    -> lineage
```

That witness can then compose with Toaster, Phonograph, Band Runtime, Haunted Camera, TranchNode, or later systems without adapter-shaped ontology.

The dependency dies; its attributable contribution remains.

## First grounded vertical slice

Do **not** begin by integrating all nine candidates.

The first implementation research slice should compose #183:

```text
VSPantry attached audio
        |
        v
Specimen Pulse research bench
        |
        +-- native/basic DSP baseline
        |
        +-- optional Essentia research reference
        |
        v
receipted comparison corpus
        |
        v
one native graduated witness
```

Recommended first graduation target: one simple, inspectable low-level witness such as transient pressure, spectral flux, or bounded energy dynamics.

Success means:

1. the reference tool can help characterize the signal;
2. all runs are specimen/version/parameter/hash bound;
3. a native eCODEsystem witness contract is written from the observed need;
4. the native implementation passes fixed differential / behavioral tests within declared tolerances;
5. the production Toaster path does not require Essentia;
6. #183 authority invariants remain unchanged;
7. final song master remains unaffected by specimen audio;
8. the native witness is reusable outside Toaster without importing Toaster-specific renderer authority.

## Research-stage dependency classes

Use three explicit buckets rather than one vague "AI dependency" category.

### GREEN — ordinary native/permissive candidate

A component may be considered for production when its license, runtime cost, determinism, packaging, privacy, and model/data provenance are acceptable.

Examples to investigate first: low-level native DSP, selected librosa-derived research, OpenCV, XState, ONNX Runtime.

### YELLOW — learned microscope

Frozen models may be used as testimony-producing instruments when exact weights/license/version/hash are known and the output remains non-authoritative.

Examples: DINOv2, MediaPipe tasks, explicitly licensed text embeddings.

### RED / RESEARCH-ONLY UNTIL CLEARED

Use as quarantined research organisms when licensing, runtime size, opacity, model terms, or dependency chains make production adoption undesirable.

Essentia is the first explicit specimen for this class.

The bucket is a project decision, not a permanent moral judgment about the upstream project.

## System-wide fit

The same witness primitive could travel beyond Haunted Toaster:

- **Haunted Toaster** — candidate diversity, motion/topology response, Specimen Pulse, typography placement, visual ecology pressure.
- **Haunted Phonograph** — rhythmic/structural seed extraction without semantic authority.
- **Band Runtime** — density, silence, collision, recurrence, shared field pressure without central assignment authority.
- **Founder Node** — neighborhood / similarity geometry without converting proximity into routing authority.
- **Corpus OS** — anomaly/pressure witnesses without statistics constituting legal state.
- **TranchNode / Project0** — deterministic projections over high-dimensional observations while projection remains witness.
- **Haunted Camera** — pose/motion/foreground/reflex sensing without a semantic AI governor.

Cross-project adoption should use BEE-style re-proving: provenance may travel; authority remains local.

## Non-goals

- no LLM supervisor;
- no autonomous tool-using agent inside Toaster;
- no AI taste ranking / automatic winner selection;
- no learned component writing canonical VisualScore or ResolvedTimeline directly;
- no ambient model inference hidden inside renderer execution;
- no cloud requirement;
- no wholesale Essentia port;
- no assumption that open-source framework license automatically covers downloaded model weights or training data;
- no production dependency added merely because it is permissively licensed;
- no claim that receipts alone extinguish copyright, patent, contract, model-license, or anticircumvention obligations.

## Acceptance criteria for this architecture slice

- [ ] The project has one canonical design checkpoint for Perceptual Reflex Layer + Compost Translation.
- [ ] #183 is named as the first grounded Toaster sensor ancestor rather than duplicated.
- [ ] Essentia is explicitly staged as First Dropout / research-only until separately cleared.
- [ ] The candidate organ catalog preserves distinct roles for librosa, OpenCV, DINOv2, MediaPipe, text embeddings, XState, ONNX Runtime, BehaviorTree.CPP, and Essentia.
- [ ] Sensor testimony is mechanically separated from proposal authority and renderer/execution authority.
- [ ] A generic sensor receipt envelope and research observation envelope are specified.
- [ ] The first future implementation target is one low-level audio witness composed with #183, not a giant model integration.
- [ ] Graduation requires a native eCODEsystem contract + fixed specimen proof + absence of the research dependency from the production path.
- [ ] Exact tool/model license/version/hash is treated as witnessed evidence, not a hard-coded timeless assumption.
- [ ] The architecture remains portable to sibling eCODEsystem projects without granting them shared execution authority.

## Stop condition

Stop this architecture slice when the project can truthfully say:

> **The eCODEsystem may borrow foreign senses, learn their useful laws, and regrow those laws in native grammar without letting the borrowed machinery become sovereign.**

Implementation should proceed later as a separate, deliberately tiny research/proof issue after this design is reviewed.

**Comment by the-static-collective** [comment]
Stewardship reconciliation: the canonical architecture checkpoint from PR #185 has now landed on `main` as `e71c5ef7b36bd6af489efa695a6a4d4de7e65384`.

This constitutes the design boundary only: sensor/foreign machinery remains testimony-only; Compost Translation does not authorize production dependency adoption; exact tool/model/version/hash/license evidence remains a future admission requirement; and #183 remains the grounded first sensor ancestor.

This issue stays open as the project-owned architecture/lineage door until a separate tiny research/proof issue is cut. No implementation, renderer authority, model stack, tag, release, or promotion is implied by the design landing.

**Comment by the-static-collective** [comment]
The promised tiny research/proof child now exists as #194.

That child carries the first bounded Compost Translation proof: one JS-native Specimen Pulse witness (transient pressure unless repository inspection proves a simpler seam), fixed synthetic probes, optional attributable foreign reference comparison, explicit authority/refusal fields, production dependency absence, and no renderer/generation widening.

With PR #185 landed on `main` and #194 owning the implementation frontier, this architecture issue has reached its stated stop condition. Closing #184 as completed; unresolved implementation evidence remains deliberately in #194/#183 rather than being implied complete here.
